### PR TITLE
fix(ui): keep markdown tooltips within the viewport

### LINKED ISSUE #16550: UI: Markdown tooltips may not fit within the viewport.
### Pre-requisites

- [x] I have double-checked my configuration
- [x] I have tested with the `:latest` image tag (i.e. `quay.io/argoproj/workflow-controller:latest`) and can confirm the issue still exists on `:latest`. If not, I have explained why, **in detail**, in my description below.
- [x] I have searched existing issues and could not find a match for this bug
- [x] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened? What did you expect to happen?

The tooltip in the Submit Workflow panel sometimes has its right edge cut off at the edge of the screen, making it unreadable.

Image

### Version(s)

v4.1.0-rc1

### Paste a minimal workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflow that uses private images.

```YAML
metadata:
  name: awesome-tiger
spec:
  entrypoint: argosay
  arguments:
    parameters:
      - name: message
        value: hello argo
        description: |-
          以下の形式でCSVを入力してください。ヘッダーも含めてパラメータに入力してください。

          email
          staff1@example.com
          staff2@example.com

          ※ 既存スタッフの最新のメールアドレスを指定した場合には、DB上の旧メールアドレスが新メールアドレスに更新されます。
  templates:
    - name: argosay
      inputs:
        parameters:
          - name: message
            value: "{{workflow.parameters.message}}"
      container:
        name: main
        image: argoproj/argosay:v2
        command:
          - /argosay
        args:
          - echo
          - "{{inputs.parameters.message}}"
```

### Logs from the workflow controller

```text
kubectl logs -n argo deploy/workflow-controller | grep ${workflow}
```

### Logs from in your workflow's wait container

```text
kubectl logs -n argo -c wait -l workflows.argoproj.io/workflow=${workflow},workflow.argoproj.io/phase!=Succeeded
```

### PR DESCRIPTION
**Pull request by panicboat**
Fixes https://github.com/argoproj/argo-workflows/issues/16550

### Motivation

It seems that using double-byte characters in tooltips causes them to be cut off at the edge of the screen, preventing them from being fully displayed.
This was an oversight in https://github.com/argoproj/argo-workflows/pull/15904.

### Modifications

- Set `boundary='viewport'` on the `ArgoTooltip` call in `ui/src/shared/components/tooltip.tsx` so Popper.js constrains the tooltip's position to the browser viewport instead of the nearest scrollable ancestor.
- Added assertions in `tooltip.test.tsx` verifying `boundary` is forwarded to `ArgoTooltip` for both markdown and non-markdown content.

### Verification

Reproduced the clipping with a minimal tippy.js harness mirroring the Submit Workflow panel's layout (a 600px-wide `SlidingPanel` docked to the right edge) and a parameter description containing a long unbroken sentence. At a 1440px-wide viewport, the tooltip's right edge measured 1565px (125px off-screen) with the default boundary; setting `boundary: 'viewport'` brought it to 1367px, fully on-screen.

```yaml
  arguments:
    parameters:
      - name: CSV_CONTENT
        default: ""
        description: |-
          以下の形式でCSVを入力してください。ヘッダーも含めてパラメータに入力してください。

          email
          staff1@example.com
          staff2@example.com

          ※ 既存スタッフの最新のメールアドレスを指定した場合には、DB上の旧メールアドレスが新メールアドレスに更新されます。
```

image

### Documentation

N/A — positioning fix only, no user-facing behavior change beyond the tooltip staying on-screen.

### AI

Used Claude Code (Anthropic) to investigate, reproduce, and fix this issue, including the regression test and this PR description.

## Summary by CodeRabbit

* **Bug Fixes**
  * Updated the Tooltip to explicitly use `boundary: 'viewport'`, keeping tooltips within the visible screen area.
* **Tests**
  * Added a unit test to verify the Tooltip forwards `boundary: 'viewport'` to the underlying Argo tooltip when given string content.

### CODE REVIEW

File: ui/src/shared/components/tooltip.tsx [hunk 1]
```diff
@@ -33,5 +33,5 @@ export function Tooltip({content, ...props}: TooltipProps) {
     ) : (
         content
     );
-    return <ArgoTooltip content={renderedContent} maxWidth={isMarkdown ? '50vw' : undefined} {...props} />;
+    return <ArgoTooltip content={renderedContent} maxWidth={isMarkdown ? '50vw' : undefined} boundary='viewport' {...props} />;
```
[the remark this answers is not in the document]
Comment (panicboat) [reply ui/src/shared/components/tooltip.tsx]: Good catch — `boundary` is indeed part of `TippyProps`, so a caller-supplied value would have won over the intended `viewport` enforcement, even though no current call site does this. Fixed in 63b249760: `{...props}` now comes first and `boundary='viewport'` is set after it, plus added a regression test asserting a caller-supplied `boundary` can't override it.

File: ui/src/shared/components/tooltip.tsx [hunk 2]
```diff
@@ -33,5 +33,5 @@ export function Tooltip({content, ...props}: TooltipProps) {
     ) : (
         content
     );
-    return <ArgoTooltip content={renderedContent} maxWidth={isMarkdown ? '50vw' : undefined} {...props} />;
+    return <ArgoTooltip content={renderedContent} maxWidth={isMarkdown ? '50vw' : undefined} {...props} boundary='viewport' />;
```
Comment (Joibel) [inline ui/src/shared/components/tooltip.tsx]: Placing `boundary='viewport'` after `{...props}` makes it non-overridable, but the exported `TooltipProps` (`React.ComponentProps<typeof ArgoTooltip>`, i.e. `TippyProps`) still advertises `boundary` as an accepted prop — so a caller writing `<Tooltip boundary={...}>` compiles cleanly and is silently ignored at runtime. It's also inconsistent with `maxWidth` on the same line, which sits *before* the spread and is caller-overridable: two opposite override semantics on one line, with nothing marking the distinction.

No call site in `ui/src` passes `boundary` today, so the fix works identically either way. Two ways to resolve the mismatch:

- If callers must never override it, encode that in the type so misuse is a compile error instead of a silent no-op:
  ```ts
  type TooltipProps = Omit<React.ComponentProps<typeof ArgoTooltip>, 'boundary'>;
  ```
- Otherwise, move it before the spread alongside `maxWidth`, matching the convention established in #15904 (and your original commit here).

Minor forward-compat note: the top-level `boundary` prop only exists in tippy v4/v5 (Popper 1); it was removed in tippy v6, where the equivalent moves into `popperOptions` and `'viewport'` is no longer a valid value. A one-line comment naming the tippy 5 dependency would keep a future argo-ui/tippy bump from turning this into a silent no-op.

**Review by Joibel — COMMENTED** [review]
### Code review

The core fix is sound: `boundary` is a real, type-checked tippy.js 5.2.1 prop, argo-ui forwards it into Tippy, and this wrapper is the only tooltip entry point in `ui/src`, so the one-line change covers every tooltip in the UI. No bugs found in the fix itself.
<sub>- If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

File: ui/src/shared/components/tooltip.test.tsx [hunk 3]
```diff
@@ -70,6 +70,7 @@ describe('Tooltip', () => {
         expect(screen.getByTestId('argo-tooltip').getAttribute('data-max-width')).toBe('50vw');
         const tippyProps = argoTooltipSpy.mock.calls[0][0];
         expect(tippyProps.onCreate).toBeUndefined();
+        expect(tippyProps.boundary).toBe('viewport');
```
Comment (Joibel) [inline ui/src/shared/components/tooltip.test.tsx]: `boundary='viewport'` is applied unconditionally in the component (unlike `maxWidth`, it isn't branched on `isMarkdown`), so the three assertions of the same fact (here, line 90, and line 99 in the dedicated test) pass or fail together — verified by mutation testing. The two piggy-backed here and at line 90 add no discriminating power beyond the dedicated test, and they sit inside tests whose names describe `maxWidth` behaviour, so a future boundary change would fail tests whose titles give no hint why. Suggest keeping only the dedicated boundary test.

File: ui/src/shared/components/tooltip.test.tsx [hunk 4]
```diff
@@ -86,6 +87,16 @@ describe('Tooltip', () => {
...
+    });
+
+    it('always enforces boundary=viewport even if a caller passes a conflicting boundary prop', () => {
+        render(
+            <Tooltip content='some string' boundary='scrollParent'>
+                <span>trigger</span>
+            </Tooltip>
+        );
+        expect(argoTooltipSpy.mock.calls[0][0].boundary).toBe('viewport');
     });
```
Comment (Joibel) [inline ui/src/shared/components/tooltip.test.tsx]: This test guards a caller that doesn't exist (no call site in `ui/src` passes `boundary`) and cements the silent-override semantics as a contract before that design question is settled — see the comment on `tooltip.tsx:36`. If the override is instead enforced in the type (`Omit<..., 'boundary'>`) or relaxed to a pre-spread default, this test gets deleted rather than adapted, so it's worth resolving that first.

To its credit, the test does have real discriminating power: moving `boundary` before the spread fails only this test. It just pins prop-spread ordering against a jest mock, not actual positioning behaviour.

File: ui/src/shared/components/tooltip.tsx (hunk 2 above)
Comment (panicboat) [reply ui/src/shared/components/tooltip.tsx]: Went with option 2 — moved `boundary='viewport'` back before the spread to match the maxWidth convention from #15904. Callers that legitimately need a different boundary can now pass one; nothing in `ui/src` does today, but the API stays symmetric with the rest of the wrapper's defaults.

Also added the tippy 5 dependency note inline as you suggested:

```tsx
// `boundary` is the tippy v5 (Popper 1) top-level API; on a v6 bump this
// moves into `popperOptions` and 'viewport' is no longer a valid value.
```

Both in 0159505d4.

File: ui/src/shared/components/tooltip.test.tsx (hunk 3 above)
Comment (panicboat) [reply ui/src/shared/components/tooltip.test.tsx]: Agreed. Dropped the piggybacked boundary assertions from both maxWidth tests (lines 73 & 90) in 0159505d4. The dedicated boundary test stays as the sole signal.

File: ui/src/shared/components/tooltip.test.tsx (hunk 4 above)
Comment (panicboat) [reply ui/src/shared/components/tooltip.test.tsx]: Yeah — with `boundary` now overridable (option 2 from your other thread), the silent-override guard is no longer the contract, so this test would be asserting the wrong thing. Replaced it in 0159505d4 with a default-boundary test whose title matches what it's actually testing:

> defaults boundary to viewport so tooltips are not clipped by scrollable ancestors

Same discriminating power as the guard version (removing `boundary='viewport'` from `tooltip.tsx` still fails it) but with a title that hints at boundary behavior and no assumption about override semantics.

**Review by Joibel — APPROVED** [review]
LGTM, thank you
